src/client/java/minicraft/entity/rideable/Boat.java

```diff
@@ -0,0 +1,292 @@
+                }
+
+                if (level.getTile(xt, yt).id != 6) {
+                    return false;
+                }
+            }
+        }
+
+        // these lists are named as if the entity has already moved-- it hasn't, though.
+        // gets all of the entities that are inside this entity (aka: colliding) before moving.
+        List<Entity> wasInside = level.getEntitiesInRect(getBounds());
+
+        int xr = 1;
+        int yr = 1;
+
+        // gets the entities that this entity will touch once moved.
+        List<Entity> isInside = level.getEntitiesInRect(new Rectangle(x + (int) xa, y + (int) ya, xr * 2, yr * 2, Rectangle.CENTER_DIMS));
+        for (int i = 0; interact && i < isInside.size(); i++) {
+            /// cycles through entities about to be touched, and calls touchedBy(this) for each of them.
+            Entity entity = isInside.get(i);
+            if (entity == this) {
+                continue; // touching yourself doesn't count XD
+            }
+
+            if (entity instanceof Player) {
+                touchedBy(entity);
+            } else {
+                touchedBy(this); // call the method. ("touch" the entity)
+            }
+        }
+
+        isInside.removeAll(wasInside); // remove all the entities that this one is already touching before moving.
+        for (Entity entity : isInside) {
+            if (entity == this) {
+                continue; // can't interact with yourself LOL
+            }
+
+            if (entity.blocks(this)) {
+                return false; // if the entity prevents this one from movement, don't move.
+            }
+        }
+
+        // finally, the entity moves!
+        x += (int) xa;
+        y += (int) ya;
+
+        return true; // the move was successful.
+    }
+
+    @Override
+    protected void touchedBy(Entity entity) {
+        if (level.getTile(this.x, this.y).id != 6) {
+            return;
+        }
+
+        if (entity instanceof Player) {
+            tryPush((Player) entity);
+        }
+    }
+
+	public void tryPush(Player player) {
+        if (level.getTile(this.x, this.y).id != 6) {
+            return;
+        }
+		if (pushTime == 0) {
+			pushDir = player.dir; // Set pushDir to the player's dir.
+			pushTime = 10; // Set pushTime to 10.
+		}
+	}
+
+	/**
+	 * "Restores" the player to the world, in other words places player out of the boat
+	 *
+	 * @param player The player in the boat
+	 */
+	public void restorePlayer(Player player) {
+		if (playerInBoat != null) {
+			playerInBoat.moveSpeed = 1.0;
+			if (playerInBoat.getLevel() == null) {
+				Game.levels[Game.currentLevel].add(player);
+				player.isRiding = false;
+				playerInBoat = null;
+			} else {
+				playerInBoat.getLevel().add(player);
+				player.isRiding = false;
+				playerInBoat = null;
+			}
+		}
+	}
+
+
+    @Override
+    public Boat clone() {
+        return new Boat();
+    }
+}
```

Comment:
This class must probably override the `die()`  method to make sure dieing works as intentional.